libyul/optimiser/UnusedAssignEliminator.cpp

```diff
@@ -74,84 +78,98 @@ void UnusedAssignEliminator::operator()(FunctionDefinition const& _functionDefin
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+		// We do not return from the current function, so it is OK to also
+		// clear the return variables.
+		m_activeStores.clear();
+}
+
 void UnusedAssignEliminator::operator()(Leave const&)
 {
 	for (YulString name: m_returnVariables)
-		changeUndecidedTo(name, State::Used);
+		markUsed(name);
+	m_activeStores.clear();
 }
 
 void UnusedAssignEliminator::operator()(Block const& _block)
 {
-	ScopedSaveAndRestore outerDeclaredVariables(m_declaredVariables, {});
-
 	UnusedStoreBase::operator()(_block);
 
-	for (auto const& var: m_declaredVariables)
-		finalize(var, State::Unused);
+	// TODO we could also move some statements from "potentially" to "toRemove".
+	for (auto const& statement: _block.statements)
+		if (auto const* varDecl = get_if<VariableDeclaration>(&statement))
+			for (auto const& var: varDecl->variables)
+				m_activeStores.erase(var.name);
 }
 
 void UnusedAssignEliminator::visit(Statement const& _statement)
 {
 	UnusedStoreBase::visit(_statement);
 
 	if (auto const* assignment = get_if<Assignment>(&_statement))
-		if (assignment->variableNames.size() == 1)
-			// Default-construct it in "Undecided" state if it does not yet exist.
-			m_stores[assignment->variableNames.front().name][&_statement];
+	{
+		// TODO this should also use user function side effects.
+		// Then we have to modify the multi-assign test (or verify that it is fine after all
+		// by adding a test where one var is used but not the other)
+		if (SideEffectsCollector{m_dialect, *assignment->value}.movable())
+		{
+			m_allStores.insert(&_statement);
+			for (auto const& var: assignment->variableNames)
+				m_activeStores[var.name] = {&_statement};
+		}
+		else
+			for (auto const& var: assignment->variableNames)
+				m_activeStores[var.name].clear();
+	}
+
+//	cerr << "After " << std::visit(AsmPrinter{}, _statement) << endl;
+//	for (auto&& [var, assigns]: m_activeStores)
+//	{
+//		cerr << "  " << var.str() << ":" << endl;
+//		for (auto const& assign: assigns)
+//			cerr << "    " << std::visit(AsmPrinter{}, *assign) << endl;
+//	}
 }
 
-void UnusedAssignEliminator::shortcutNestedLoop(TrackedStores const& _zeroRuns)
+void UnusedAssignEliminator::shortcutNestedLoop(ActiveStores const& _zeroRuns)
 {
 	// Shortcut to avoid horrible runtime:
 	// Change all assignments that were newly introduced in the for loop to "used".
 	// We do not have to do that with the "break" or "continue" paths, because
 	// they will be joined later anyway.
 	// TODO parallel traversal might be more efficient here.
-	for (auto& [variable, stores]: m_stores)
+
+	// TODO is this correct?
+
+	for (auto& [variable, stores]: m_activeStores)
 		for (auto& assignment: stores)
 		{
 			auto zeroIt = _zeroRuns.find(variable);
```

Comment:
```suggestion
		auto zeroIt = _zeroRuns.find(variable);
		for (auto& assignment: stores)
		{
```
This should be more efficient, no?